### PR TITLE
Enhancing the FTP protocol

### PR DESCRIPTION
**Pull request by RomanRII**
# What was done
* Modified `nxc/protocols/ftp.py`
  * Modified the `--ls` flag to allow for listing the current directory and sub-directories. Default now lists `.`. If an argument is provided, it will list the provided sub-directory
  * Added the `--get` flag to download a file on the server. If the file exists and is successfully downloaded, it will be written to the users cwd with the remote file's filename.
  * Added the `--put` flag to upload files onto the server. 
* Modified `nxc/protocols/ftp/proto_args.py` to reflect the added features
  * Modified the `--ls` flag to allow for a default directory listing (`.`) or use a provided directory
  * Added the `--get` and `--put` flags
* Modified [nxc/protocols/ftp.py#L83](https://github.com/Pennyw0rth/NetExec/blob/main/nxc/protocols/ftp.py#L83) to comply with [RFC 1635](https://www.rfc-editor.org/rfc/rfc1635.html)

# Test environment used
* During testing, different permissions were used to account for edge cases.
``` python3
from pyftpdlib.authorizers import DummyAuthorizer
from pyftpdlib.handlers import FTPHandler
from pyftpdlib.servers import FTPServer

def start_ftp_server():
    # Instantiate an authorizer object responsible for managing authentication and authorization
    authorizer = DummyAuthorizer()

    # Allow anonymous access with read-only permissions
    authorizer.add_anonymous("./FTP_Folder/", perm="erlw")

    # Instantiate an FTP handler object and link it to the authorizer
    handler = FTPHandler
    handler.authorizer = authorizer

    # Create the FTP server on the desired address and port, and link it to the handler
    server = FTPServer(("127.0.0.1", 21), handler)

    # Start the FTP server
    server.serve_forever()

if __name__ == "__main__":
    start_ftp_server()
```

### CODE REVIEW

**Comment by RomanRII** [comment]
# RFC 1635 Compliance
image

# --ls 
## Standard usage
image
## If permissions are not granted
image

# --get
## Standard usage
image
## If the file does not exist
image

# --put
## Standard usage
image
## If a local file does not exist
image
## Permissions error
image

**Comment by RomanRII** [comment]
Couple things to mind about this merge, it currently doesn't allow the following:
## --put
* If a user attempts to upload a file like `--put example.txt .`, it will fail with `Failed to upload file. Response: (550 Is a directory.)`

The user will need to specify the filename or dir + filename to save to
`--put example.txt example.txt` or `--put example.txt Sub/Dir/example.txt`

If this current implementation is not prefered, let me know and I can make some adjustments to allow users to upload using either `.` or `Sub/` and upload to the directory using the local_name

**Comment by Marshall-Hallenbeck** [comment]
@RomanRII can you add these new flags to the e2e tests commands file under the FTP section (tests/e2e_commands.txt)?

**Comment by RomanRII** [comment]
@Marshall-Hallenbeck you got it. will get that done

**Comment by RomanRII** [comment]
* Added new flags to ([tests/e2e_commands.txt](https://github.com/RomanRII/NetExec/blob/main/tests/e2e_commands.txt#L207))
* Added test file to test upload/download functions ([tests/data/test_file.txt](https://github.com/RomanRII/NetExec/blob/main/tests/data/test_file.txt))

File: nxc/protocols/ftp/proto_args.py [hunk 1]
```diff
@@ -3,5 +3,7 @@ def proto_args(parser, std_parser, module_parser):
     ftp_parser.add_argument("--port", type=int, default=21, help="FTP port (default: 21)")
 
     cgroup = ftp_parser.add_argument_group("FTP Access", "Options for enumerating your access")
-    cgroup.add_argument("--ls", action="store_true", help="List files in the directory")
+    cgroup.add_argument("--ls", metavar="DIRECTORY", nargs="?", const=".", help="List files in the directory")
+    cgroup.add_argument("--get", metavar="FILE", help="Download a file")
+    cgroup.add_argument("--put", metavar=("LOCAL_FILE", "REMOTE_FILE"), nargs=2, help="Upload a file")
```
Comment (NeffIsBack) [inline nxc/protocols/ftp/proto_args.py:8]: Would be nice if you could add an example as in:
https://github.com/Pennyw0rth/NetExec/blob/66e6c95f0d4cabe4acabdfadaa69f8eb38bb056f/nxc/protocols/smb/proto_args.py#L75C45-L75C45

File: nxc/protocols/ftp.py [hunk 2]
```diff
@@ -80,16 +81,42 @@ def plaintext_login(self, username, password):
             host_id = self.db.get_hosts(self.host)[0].id
             self.db.add_loggedin_relation(cred_id, host_id)
 
-            if username in ["anonymous", ""] and password in ["", "-"]:
+            if username in ["anonymous", ""]:
                 self.logger.success(f"{username}:{process_secret(password)} {highlight('- Anonymous Login!')}")
             else:
                 self.logger.success(f"{username}:{process_secret(password)}")
-
+
```
Comment (NeffIsBack) [inline nxc/protocols/ftp.py:88]: Please remove the space

**Review by NeffIsBack — CHANGES_REQUESTED** [review]
Really good PR! Clean code and very nice documentation of the changes! Love to see that

File: nxc/protocols/ftp/proto_args.py (hunk 1 above)
Comment (RomanRII) [reply nxc/protocols/ftp/proto_args.py]: Added: https://github.com/RomanRII/NetExec/commit/901b8dee80fc1109414c688f491f5ad4fdec9049

File: nxc/protocols/ftp.py (hunk 2 above)
Comment (RomanRII) [reply nxc/protocols/ftp.py]: Removed space: https://github.com/RomanRII/NetExec/commit/f26b676c8ee94992006352737663352301ed201b

**Review by Marshall-Hallenbeck — APPROVED** [review]
Reviewed and tested this. It all works. Great code and feature, thank you!

**Review by NeffIsBack — APPROVED** [review]
